fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt

```diff
@@ -0,0 +1,67 @@
+package org.mozilla.fenix.shopping.middleware
+
+import mozilla.components.browser.engine.gecko.shopping.GeckoProductAnalysis
+import mozilla.components.browser.engine.gecko.shopping.Highlight
+import mozilla.components.concept.engine.shopping.ProductAnalysis
+import org.mozilla.fenix.shopping.store.ReviewQualityCheckState
+import org.mozilla.fenix.shopping.store.ReviewQualityCheckState.OptedIn.ProductReviewState
+
+/**
+ * Maps [ProductAnalysis] to [ProductReviewState].
+ */
+fun ProductAnalysis?.toProductReviewState(): ProductReviewState =
+    if (this == null) {
+        ProductReviewState.Error
+    } else {
+        when (this) {
+            is GeckoProductAnalysis -> toProductReview()
+            else -> ProductReviewState.Error
+        }
+    }
+
+private fun GeckoProductAnalysis.toProductReview(): ProductReviewState =
+    if (productId == null) {
+        ProductReviewState.Error
+    } else {
+        val mappedRating = adjustedRating.toFloatOrNull()
+        val mappedGrade = grade?.toGrade()
+        val mappedHighlights = highlights?.toHighlights()?.toSortedMap()
+
+        if (mappedGrade == null && mappedRating == null && mappedHighlights == null) {
+            ProductReviewState.NoAnalysisPresent
+        } else {
+            ProductReviewState.AnalysisPresent(
+                productId = productId!!,
+                reviewGrade = mappedGrade,
+                needsAnalysis = needsAnalysis,
+                adjustedRating = mappedRating,
+                productUrl = analysisURL!!,
+                highlights = mappedHighlights,
+            )
+        }
+    }
+
+private fun String.toGrade(): ReviewQualityCheckState.Grade? =
+    try {
+        ReviewQualityCheckState.Grade.valueOf(this)
+    } catch (e: IllegalArgumentException) {
+        null
+    }
+
+private fun Highlight.toHighlights(): Map<ReviewQualityCheckState.HighlightType, List<String>>? {
+    val highlights: Map<ReviewQualityCheckState.HighlightType, List<String>?> = mapOf(
+        ReviewQualityCheckState.HighlightType.QUALITY to quality,
+        ReviewQualityCheckState.HighlightType.PRICE to price,
+        ReviewQualityCheckState.HighlightType.SHIPPING to shipping,
+        ReviewQualityCheckState.HighlightType.COMPETITIVENESS to competitiveness,
+    )
+
+    return highlights.filterValues { it != null }.mapValues { it.value!! }.ifEmpty { null }
+}
+
+private fun Double.toFloatOrNull(): Float? =
+    if (this == 0.0) {
```

Comment:
is this implying the service will never return 0.0?

Reply:
Yes, my understanding is that rated products would have at least 1 as value, since people can't rate products as 0. 
Gecko layer is converting null values to `0.0` so till it's changed there back to null, we have to add this check unfortunately.

Reply:
might be worth doc'ing the 'why' here